### DISCUSSION TITLE
Ideal way to pass data to modals in AlpineJS?

### DISCUSSION THREAD

**Discussion by bmarsh9** [alpinejs/alpine#3556] [1. Help]
Frequently I am showing/hiding modals on the screen and passing data to the modal. Before I continue with the pattern below, I'd like to make sure I'm doing it the best "AlpineJS" way. Here is the code for what I am doing - basically a listing of names, when a name is clicked, we spawn the modal and load the data into the modal. So my question is, is this the ideal way to achieve this in AlpineJS? Or should I be doing it another way?

```
<!-- html for showing a list of names from the items array -->
<div x-data=app()>
  <template x-for="(item, index) in items">
    <p :click="dictForActiveModal=item;modalActive=true" x-text="item.name"></p>
  </template>
</div>
<!-- end array listing -->

<!-- hidden modal that is shown/hidden and we want to populate with data from dictForActiveModal -->
<modal x-show="modalActive">
  <p x-text="dictForActiveModal.name"></p>
</modal>
<!-- end modal -->

<script>
  function app() {
    return {
      modalActive = false,
      dictForActiveModal = {},
      items = [{"name":"tom"}, {"name:"beth"}],
      fetchItems: function() {
        // set the items array from server
      }
    }
  }
</script>
```

**Comment by ekwoka** [comment]
I am partial to directly putting the modal inside the scope of the loop with an x-if so each one can control it's own modal scope.

```html
<template x-for"item in items" :key="item.id">
  <div>
    <div x-data="{open:false}">
      <button @click="open = !open">
        show
      </button>
      <template x-if="open">
        <modal />
      </template>
    </div>
  </div>
</template>
```

Also, don't put click listeners on non-focusable tags 😄

**Comment by dfallman** [comment]
For me, that ultimately depends on what you're building and how much you need the modal to know. If you have modals (or simpler elements like dropdowns) that don't need to know anything about the state of the rest of the app then the local `x-data="open:false"` is great. 

But if you have an app that needs to respond to other things than direct user input -- for instance if the app should be able to modify the modal, or the modal/element needs data from the app, or if the modal/element needs to know if other modal/elements(s) are currently open, etc. -- then something like `<button @click.stop="modalActive=item;isModalActive=true" x-text="item.name"></button>` seems the most straightforward way. 

We're using a similar scheme with some helper functions in our project, so that we show a modal with something like `showModal('modal1')` and `isModalOpen()` etc. But then again in our app most of the work is done in a single alpine context. 

You could also use `Alpine.store()` as well as `$dispatch` to do similar things.
